server/src-lib/Hasura/Eventing/EventTrigger.hs

```diff
@@ -158,45 +158,56 @@ initEventEngineCtx maxT _eeCtxFetchInterval = do
-  :: (HasVersion) => L.Logger L.Hasura -> LogEnvHeaders -> HTTP.Manager-> Q.PGPool
-  -> IO SchemaCache -> EventEngineCtx -> LockedEventsCtx
-  -> IO void
-processEventQueue logger logenv httpMgr pool getSchemaCache eeCtx@EventEngineCtx{..} LockedEventsCtx {leEvents}= do
+  :: forall m void
+   . ( HasVersion
+     , MonadIO m
+     , MonadBaseControl IO m
+     , LA.Forall (LA.Pure m)
+     , MonadMask m
+     )
+  => L.Logger L.Hasura
+  -> LogEnvHeaders
+  -> HTTP.Manager
+  -> Q.PGPool
+  -> IO SchemaCache
+  -> EventEngineCtx
+  -> LockedEventsCtx
+  -> m void
+processEventQueue logger logenv httpMgr pool getSchemaCache eeCtx@EventEngineCtx{..} LockedEventsCtx{leEvents} = do
   events0 <- popEventsBatch
   go events0 0 False
   where
     fetchBatchSize = 100
     popEventsBatch = do
-      let run = runExceptT . Q.runTx pool (Q.RepeatableRead, Just Q.ReadWrite)
+      let run = liftIO . runExceptT . Q.runTx pool (Q.RepeatableRead, Just Q.ReadWrite)
       run (fetchEvents fetchBatchSize) >>= \case
           Left err -> do
-            L.unLogger logger $ EventInternalErr err
+            liftIO $ L.unLogger logger $ EventInternalErr err
             return []
           Right events -> do
             saveLockedEvents (map eId events) leEvents
             return events
 
     -- work on this batch of events while prefetching the next. Recurse after we've forked workers
     -- for each in the batch, minding the requested pool size.
-    go :: [Event] -> Int -> Bool -> IO void
+    go :: [Event] -> Int -> Bool -> m void
     go events !fullFetchCount !alreadyWarned = do
       -- process events ASAP until we've caught up; only then can we sleep
-      when (null events) $ sleep _eeCtxFetchInterval
+      when (null events) . liftIO $ sleep _eeCtxFetchInterval
 
       -- Prefetch next events payload while concurrently working through our current batch.
       -- NOTE: we probably don't need to prefetch so early, but probably not
       -- worth the effort for something more fine-tuned
-      eventsNext <- withAsync popEventsBatch $ \eventsNextA -> do
+      eventsNext <- LA.withAsync popEventsBatch $ \eventsNextA -> do
         -- process approximately in order, minding HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE:
         forM_ events $ \event -> do
-             t <- async $ runReaderT (withEventEngineCtx eeCtx $ (processEvent event)) (logger, httpMgr)
-             -- removing an event from the _eeCtxLockedEvents after the event has
-             -- been processed
-             removeEventFromLockedEvents (eId event) leEvents
-             link t
-
-        -- return when next batch ready; some 'processEvent' threads may be running.
-        wait eventsNextA
+          processEvent event
+            & withEventEngineCtx eeCtx
+            & flip runReaderT (logger, httpMgr)
+          -- removing an event from the _eeCtxLockedEvents after the event has
+          -- been processed
+          removeEventFromLockedEvents (eId event) leEvents
```

Comment:
looks like the `link t` part is missing?

Reply:
Good catch... Strange that that got removed.

Reply:
Ah actually, is there a thread to be linked?

Reply:
Seems like this action is no longer performed async, so there's nothing to link.

Reply:
This was a regression in an earlier master. `processEvent` needs to be async.

Reply:
Ah ok. I'll try patching this.

Reply:
I've put in an async call and link again:


```
      eventsNext <- LA.withAsync popEventsBatch $ \eventsNextA -> do
        -- process approximately in order, minding HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE:
        forM_ events $ \event -> do
          t <- processEvent event
            & withEventEngineCtx eeCtx
            & flip runReaderT (logger, httpMgr)
            & LA.async
          -- removing an event from the _eeCtxLockedEvents after the event has
          -- been processed
          removeEventFromLockedEvents (eId event) leEvents
          LA.link t
        LA.wait eventsNextA
```